.gitlab-ci.yml

```diff
@@ -1,97 +1,213 @@
+  variables:
+    PUPPET_VERSION: '~> 4.8.0'
+  <<: *cache_bundler
+  <<: *setup_bundler_env
+  <<: *validation_checks
+
+pup4.8-unit:
+  stage: unit
+  tags:
     - docker
-  image: ruby:2.3
-  allow_failure: true
-  script: 
-    - bundle install
-    - bundle exec rake spec
+  image: ruby:2.1
+  variables:
+    PUPPET_VERSION: '~> 4.8.0'
+  <<: *cache_bundler
+  <<: *setup_bundler_env
+  <<: *spec_tests
+
 
-# For PE LTS Support
+# Puppet 4.10 for PE 2017.2 support (EOL:2018-02-21)
 # See: https://puppet.com/misc/puppet-enterprise-lifecycle
-puppet4.7-syntax: 
-  stage: test
-  tags: 
+# --------------------------------------
+pup4.10-validation:
+  stage: validation
+  tags:
     - docker
   image: ruby:2.1
-  script: 
-    - PUPPET_VERSION=4.7 bundle install
-    - PUPPET_VERSION=4.7 bundle exec rake syntax
-unit-test-puppet4.7-ruby-2.1:
-  stage: test
-  tags: 
+  variables:
+    PUPPET_VERSION: '~> 4.10.0'
+  <<: *cache_bundler
+  <<: *setup_bundler_env
+  <<: *validation_checks
+
+pup4.10-unit:
+  stage: unit
+  tags:
     - docker
   image: ruby:2.1
-  script: 
-    - PUPPET_VERSION=4.7 bundle install
-    - PUPPET_VERSION=4.7 bundle exec rake spec
-unit-test-puppet4.7-ruby-2.2:
-  stage: test
-  tags: 
+  variables:
+    PUPPET_VERSION: '~> 4.10.0'
+  <<: *cache_bundler
+  <<: *setup_bundler_env
+  <<: *spec_tests
+
+
+# Puppet 5.3 for PE 2017.3 support (EOL: 2018-07)
+# See: https://puppet.com/misc/puppet-enterprise-lifecycle
+# --------------------------------------
+pup5.3-validation:
+  stage: validation
+  tags:
     - docker
-  image: ruby:2.2
+  image: ruby:2.4
+  variables:
+    PUPPET_VERSION: '~> 5.3.0'
+  <<: *cache_bundler
+  <<: *setup_bundler_env
+  <<: *validation_checks
+
+pup5.3-unit:
+  stage: unit
+  tags:
+    - docker
+  image: ruby:2.4
+  variables:
+    PUPPET_VERSION: '~> 5.3.0'
+  <<: *cache_bundler
+  <<: *setup_bundler_env
+  <<: *spec_tests
   allow_failure: true
-  script: 
-    - PUPPET_VERSION=4.7 bundle install
-    - PUPPET_VERSION=4.7 bundle exec rake spec
-unit-test-puppet4.7-ruby-2.3:
-  stage: test
-  tags: 
+
+
+# Keep an eye on the latest puppet 5
+# ----------------------------------
+pup5.latest-validation:
+  stage: validation
+  tags:
     - docker
-  image: ruby:2.3
+  image: ruby:2.4
+  variables:
+    PUPPET_VERSION: '~> 5.0'
+  <<: *cache_bundler
+  <<: *setup_bundler_env
+  <<: *validation_checks
   allow_failure: true
-  script: 
-    - PUPPET_VERSION=4.7 bundle install
-    - PUPPET_VERSION=4.7 bundle exec rake spec
 
-acceptance-test: 
-  stage: test
-  tags: 
+pup5.latest-unit:
+  stage: unit
+  tags:
+    - docker
+  image: ruby:2.4
+  variables:
+    PUPPET_VERSION: '~> 5.0'
+  <<: *cache_bundler
+  <<: *setup_bundler_env
+  <<: *spec_tests
+  allow_failure: true
+
+
+
+# Acceptance tests
+# ==============================================================================
+default:
+  stage: acceptance
+  tags:
+    - beaker
+  <<: *cache_bundler
+  <<: *setup_bundler_env
```

Comment:
Don't run `gem install` on acceptance tests

Reply:
it's already installed

Reply:
@jeefberkey Are you assuming that on the basis of the cache loading from previous jobs in the pipeline?  There's no guarantee that the Runner cache hasn't been cleared (or borked) since the pipeline began.  The job might be running as a "retry" long after the original pipeline, too.

Or perhaps your concerns are driven by the performance hit of running `gem install` for each job?  We could try to optimize that by running something more nuanced like `gem install  --local bundler ARGS || gem install bundler ARGS `

Either way, I think we'll have to be prepared to run _something_ to ensure the environment is prepared correctly.

Reply:
@jeefberkey I've created [SIMP-4520](https://simp-project.atlassian.net/browse/SIMP-4520) to track related improvements to the Gitlab CI pipeline.

Reply:
Concur with @op-ct since I've started seeing things randomly fail with weird Ruby errors that are fixed when the cache is cleared.